metrics/config.go

```diff
@@ -0,0 +1,200 @@
+	}
+	logger.Info("Created Opencensus Stackdriver exporter with config:", config)
+	return e, nil
+}
+
+func newPrometheusExporter(config metricsConfig, logger *zap.SugaredLogger) (view.Exporter, error) {
+	e, err := prometheus.NewExporter(prometheus.Options{Namespace: config.component})
+	if err != nil {
+		logger.Error("Failed to create the Prometheus exporter.", zap.Error(err))
+		return nil, err
+	}
+	logger.Info("Created Opencensus Prometheus exporter with config:", config)
+	logger.Info("Start the endpoint for Prometheus exporter.")
+	// Start the endpoint for Prometheus scraping
+	go func() {
+		sm := http.NewServeMux()
+		sm.Handle("/metrics", e)
+		promSrv := &http.Server{
+			Addr:    ":9090",
+			Handler: sm,
+		}
+		promSrvChan <- promSrv
+		promSrv.ListenAndServe()
+	}()
+	return e, nil
+}
+
+func getMetricsConfig(m map[string]string, domain string, component string, logger *zap.SugaredLogger) (metricsConfig, error) {
+	var mc metricsConfig
+	backend, ok := m[backendDestinationKey]
+	if !ok {
+		return mc, errors.New("metrics.backend-destination key is missing")
+	}
+	var mb MetricsBackend
+	if strings.EqualFold(backend, "stackdriver") {
+		mb = Stackdriver
+	} else {
+		mb = Prometheus
+		if !strings.EqualFold(backend, "prometheus") {
+			logger.Infof("Unsupported metrics backend value \"%s\". Use the default metrics backend prometheus.", backend)
+		}
+	}
+	mc.backendDestination = mb
+
+	sdProj, ok := m[stackdriverProjectIDKey]
+	if strings.EqualFold(backend, "stackdriver") && !ok {
+		return mc, errors.New("metrics.stackdriver-project-id key is missing when the backend-destination is set to stackdriver.")
+	}
+	if strings.EqualFold(backend, "stackdriver") && sdProj == "" {
+		return mc, errors.New("metrics.stackdriver-project-id field cannot be empty.")
+	}
+	mc.stackdriverProjectID = sdProj
+
+	if domain == "" {
+		logger.Info("Metrics domain name missing. Use \"domain\"")
+		mc.domain = "domain"
+	} else {
+		mc.domain = domain
+	}
+
+	if component == "" {
+		logger.Info("Metrics component name missing. Use \"component\"")
+		mc.component = "component"
+	} else {
+		mc.component = component
+	}
+	return mc, nil
+}
+
+// UpdateExporterFromConfigMap returns a helper func that can be used to update the exporter
+// when a config map is updated
+func UpdateExporterFromConfigMap(domain string, component string, logger *zap.SugaredLogger) func(configMap *corev1.ConfigMap) {
+	return func(configMap *corev1.ConfigMap) {
+		newConfig, err := getMetricsConfig(configMap.Data, domain, component, logger)
+		if err != nil {
+			if exporter == nil {
+				// Fail the process if there doesn't exist an exporter.
+				logger.Fatal("Failed to get a valid metrics config")
+			} else {
+				logger.Error("Failed to get a valid metrics config; Skip updating the metrics exporter", zap.Error(err))
+				return
+			}
+		}
+		if newConfig.backendDestination != mConfig.backendDestination ||
+			newConfig.stackdriverProjectID != mConfig.stackdriverProjectID {
```

Comment:
If destination is prometheus, but sdprojectid changes, this will recreate the exporter, which is not needed. I would restructure this for easier reading:
```
changed bool
if newConfig.backendDestination != mConfig.backendDestination {
    changed = true
} else if newConfig.backendDestination == stackdriver && newConfig.stackdriverProjectID != mConfig.stackdriverProjectID) {
    changed = true
}

if changed {
   err = newMetricsExporter...
...
}
```